awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx

```diff
@@ -1,94 +1,688 @@
+    const invalidLinkTargetIds = [];
+    // Find and mark any ancestors as disabled to prevent cycles
+    links.forEach(link => {
+      // id=1 is our artificial root node so we don't care about that
+      if (link.source.id !== 1) {
+        if (link.source.id === sourceNode.id) {
+          // Disables direct children from the add link process
+          invalidLinkTargetIds.push(link.target.id);
+        }
+        if (!parentMap[link.target.id]) {
+          parentMap[link.target.id] = [];
+        }
+        parentMap[link.target.id].push(link.source.id);
+      }
+    });
+
+    const getAncestors = id => {
+      if (parentMap[id]) {
+        parentMap[id].forEach(parentId => {
+          invalidLinkTargetIds.push(parentId);
+          getAncestors(parentId);
+        });
+      }
+    };
+
+    getAncestors(sourceNode.id);
+
+    // Filter out the duplicates
+    invalidLinkTargetIds
+      .filter((element, index, array) => index === array.indexOf(element))
+      .forEach(ancestorId => {
+        newNodes.forEach(node => {
+          if (node.id === ancestorId) {
+            node.isInvalidLinkTarget = true;
+          }
+        });
+      });
+
+    setAddLinkSourceNode(sourceNode);
+    setAddingLink(true);
+    setNodes(newNodes);
+  };
+
+  const selectTargetNodeForLinking = targetNode => {
+    setAddLinkTargetNode(targetNode);
+  };
+
+  const addLink = linkType => {
+    const newLinks = [...links];
+    const newNodes = [...nodes];
+
+    newNodes.forEach(node => {
+      node.isInvalidLinkTarget = false;
+    });
+
+    newLinks.push({
+      source: { id: addLinkSourceNode.id },
+      target: { id: addLinkTargetNode.id },
+      linkType,
+      type: 'link',
+    });
+
+    newLinks.forEach((link, index) => {
+      if (link.source.id === 1 && link.target.id === addLinkTargetNode.id) {
+        newLinks.splice(index, 1);
+      }
+    });
+
+    if (!unsavedChanges) {
+      setUnsavedChanges(true);
+    }
+    setAddLinkSourceNode(null);
+    setAddLinkTargetNode(null);
+    setAddingLink(false);
+    setLinks(newLinks);
+  };
+
+  const cancelNodeLink = () => {
+    const newNodes = [...nodes];
+
+    newNodes.forEach(node => {
+      node.isInvalidLinkTarget = false;
+    });
+
+    setAddLinkSourceNode(null);
+    setAddLinkTargetNode(null);
+    setAddingLink(false);
+    setNodes(newNodes);
+  };
+
+  const deleteAllNodes = () => {
+    setAddLinkSourceNode(null);
+    setAddLinkTargetNode(null);
+    setAddingLink(false);
+    setNodes(
+      nodes.map(node => {
+        if (node.id !== 1) {
+          node.isDeleted = true;
+        }
+
+        return node;
+      })
+    );
+    setLinks([]);
+    setShowDeleteAllNodesModal(false);
+  };
+
+  const handleVisualizerClose = () => {
+    if (unsavedChanges) {
+      setShowUnsavedChangesModal(true);
+    } else {
+      history.push(`/templates/workflow_job_template/${template.id}/details`);
+    }
+  };
+
+  const handleVisualizerSave = async () => {
```

Comment:
450 line function. This would be much easier to follow once it's broken up into smaller pieces. Breaking it up should also help simplify a lot of the complex/nested if-else logic in the first section.

For example, down near the end, lines 518–650 could be simplified to this, with the looping & bare logic relegated to smaller functions:

```
const {linkMap, newLinks} = buildNewLinkMap(links, originalLinkMap);
const disassociateRequests = disassociateDeletedLinks(linkMap, originalLinkMap, deletedNodeIds);
await Promise.all(disassociateRequests);

const associateRequests = associateNewLinks(newLinks);
await Promise.all(associateRequests);
history.push(`/templates/workflow_job_template/${template.id}/details`);
```

If you apply this sort of treatment to the whole function, I think you could probably get it down to maybe 50 lines or so. Then it's much easier to understand at a glance what's going on in the saving process.